### ISSUE TITLE
Implement code overrides matching via tokens parsing

### ISSUE THREAD

**Issue by krassowski** [krassowski/jupyterlab-lsp#347]
## What are you trying to do?

Add token based code overrides.

## How is it done today, and what are the limits of current practice?

We only support regular expression based overrides.

## Who cares? If you are successful, what difference will it make?

- screening code blocks with multiple complex regular expressions is expensive
- regular expressions cannot distinguish escaped code (not enough memory, fundamental grammar limitation)

With good token parser we can hugely reduce the computational complexity.

**Comment by blois** [comment]
Are there languages other than Python that have cell magics? This almost feels like an implementation detail of the LSP that it should be supporting 'Notebook Python' but the underlying implementation only supports Python.

Could cell magics be implemented via an extension to the existing Python LSP engines?

> will kernel developers in Jupyter community be willing to go this route, or would they prefer that a LSP server is implemented for each kernel / existing language servers are amended for each kernel?

I'm of the (possibly unpopular) opinion that magics introduce significant unexpected tooling overhead. This includes lsp, syntax highlighting, lint, format and graduating code to files). Languages should think very hard before adding them.

**Comment by krassowski** [comment]
Thank you for chiming in @blois. I respect your point of view, and agree that if there were no magics, the overhead in maintaining the tooling would not be as high.

However, as a happy user of IPython I could not see it without magics; I use them heavily everyday and I believe they are a huge contributor to the IPython success as compared to some other kernels that explicitly decided not to support magics (excluding kernels of languages which do not need magics as the syntax of the language is very flexible in itself).

> Could cell magics be implemented via an extension to the existing Python LSP engines?

No, I believe that implementing magics in LSP servers would not be realistically possible. In addition to the reasons listed in _Why should it be declared by the kernel? Should not the LSP servers work harder to adopt to the language variation?_ ([above](https://github.com/krassowski/jupyterlab-lsp/issues/347#issue-697309131)):
- there are dozens of LSP servers for each language. A Python user might like pyls today as it is open source, but tomorrow want to switch to python-language-server from Microsoft, and a day after that to pylance (also from MS). The last one is not open source, but probably the most advanced one. How would you implement magics into a closed-source server?
- there are dozens of kernels which use similar base magics for different languages; the LSP server can be written in these languages and I just cannot imagine how one would go implementing magics in each of these servers to cater specialised kernels with user base in thousands rather than millions. In my view these kernels are a vital part of the Jupyter ecosystem which is open to to diversity and niche needs required in specialised areas.
   - the whole point of having [Metakernel](https://github.com/Calysto/metakernel), or now xeus was to make creating kernels easy for anyone
   - the whole point of LSP is to solve `n:m` problem (`n` editors need to adapt to support `m` languages)

> Are there languages other than Python that have cell magics

Kernels that support magics in general:
- [IScala](https://github.com/mattpap/IScala#magics) supports magics
- [IPolyglot](https://github.com/hpi-swa/ipolyglot) is build around cell magics; magics are essential to the existence of this kernel
- [Perl 6/Raku kernel](https://github.com/bduggan/p6-jupyter-kernel) kernel supports magics (`#%` and `%%` syntax)
- [coalac](https://github.com/sagemathinc/cocalc/wiki/sagews-custom-modes) and embedded SageMath kernels extensively support magics, but renamed them to "modes" to [make it easier to understand for users](https://github.com/sagemathinc/cocalc/issues/722) (I don't think it is easier for non-native speakers but that's a fair choice)
- [IPyStata](https://github.com/TiesdeKok/ipystata) is build around cell magics; magics are essential to the existence of this kernel
- All kernels based on [Metakernel](https://github.com/Calysto/metakernel) support magics:
  > - Basic set of line and cell magics for all kernels.
  >    - Python magic for accessing python interpreter.
  >    - Run kernels in parallel.
  >    - Shell magics.
  >    - Classroom management magics.
  > - Tab completion for magics and file paths.
  > - Help for magics using ? or Shift+Tab.
  > - Plot magic for setting default plot behavior.
  - this includes: [matlab_kernel](https://github.com/Calysto/matlab_kernel#additional-information), [octave_kernel](https://github.com/Calysto/octave_kernel#usage), [calysto_processing kernel](https://github.com/Calysto/calysto_processing) and _dozens_ more.

Kernels that do not support magics:
- IJulia - [tolerates but does not support](https://julialang.github.io/IJulia.jl/stable/manual/usage/#Julia-and-IPython-Magics) (i.e. IPython magics are essentially help commands to make things easier to users switching over from IPython)
- [IRKernel](https://irkernel.github.io/faq/) - a design decision, though users repeatedly asked for magics support
- [IElixir](https://github.com/pprzetacznik/IElixir#magic-commands) only supports magical namespace population (which can be addressed by token parsing as we can introduce 'cell-end' token with value specifying cell number); this is akin to IPython `_` variables.
- [IRuby](https://github.com/SciRuby/iruby/pull/140#issuecomment-529923243) does not support and did not plan on supporting magics as of September 2019
- [IHaskell](https://github.com/gibiansky/IHaskell/wiki#ihaskell-directives) supports ihaskell-directives, effectively being line magics equivalents; they do not support defining custom magics, cell magics not shell expressions (as those are kind-of easy to get with command-qq), but they are [open to consider cell magics support in the future](https://github.com/gibiansky/IHaskell/issues/77#issuecomment-31613715).

I did not look _specifically_ for _cell_ magics - why are you focusing on cell magics only?

**Comment by blois** [comment]
I agree on the usefulness of magics, really trying to figure out a way to break the problem down some.

For moving support into the LSP- I wonder if it could be done with one LSP wrapping another and doing the translation at that level. Some of the implementation would be tricky but I'm not sure it's much different than what needs to be done at the editor level. It sounds like VSCode may be trying something along these lines for their notebook support? I see no mention of supporting magics there though.
